### ISSUE TITLE
Nested object validation assumes existence of optional parent nodes

### ISSUE THREAD

**Issue by kgryte** [mafintosh/is-my-json-valid#14]
Currently using IMJV to create a schema [validator](https://github.com/figure-io/matrix-diagram-spec). Ran into a few failing tests which seem to be related to the same issue. 

``` javascript
    if (data.x !== undefined) {
        if (!(typeof data.x === "object" && data.x && !Array.isArray(data.x))) {
          errors++
          if (validate.errors === null) validate.errors = []
          validate.errors.push(error35)
        } else {
          if (data.x.y.z === undefined) {
            errors++
            if (validate.errors === null) validate.errors = []
            validate.errors.push(error36)
          } else {
            if (!(typeof data.x.y === "object" && data.x.y && !Array.isArray(data.x.y))) {
              errors++
              if (validate.errors === null) validate.errors = []
              validate.errors.push(error37)
            } else {
              if (data.x.y.z !== undefined) {
                if (!(typeof data.x.y.z === "string")) {
                  errors++
                  if (validate.errors === null) validate.errors = []
                  validate.errors.push(error38)
                } else {
                }
              }
            }
          }
        }
      }
   }
```

The issue arises when a nested object is an optional property of a parent node. In the above example, `x` is an optional property of `data`. The generated validator correctly checks to make sure the `x` property exists before continuing with validation. 

Similarly, `y` is an optional property of the object `x`. If `y` exists, it is also an object having properties; e.g., `z`. In the above code, after confirming that `data.x` is an object, the validator proceeds to determine if `data.x.y.z` exists __before__ having determined that `data.x.y` exists.

This causes problems. For example, `data.x = {}` will throw, as you cannot read property `z` of `undefined`.

To my knowledge, what should happen is the following:

1. Check if optional property `x` exists.
  - If yes, proceed to 2.
2. Check if `data.x` is an `object`.
  - If no, return an error; else proceed to step 3.
3. Check if optional property `y` exists.
  - If no, return to step 1 and check for next property; else proceed to step 4.
4. Check if `data.x.y` is an `object`.
  - If no, return an error; else proceed to step 5.
5. Check if required property `z` exists.
  - If no, return an error; else proceed to step 6.
6. Check if `data.x.y.z` is a `string`.
  - If no, return an error; else return to step 1 and check for next property.

If so, then the code generation needs to be altered to reflect this sequence.

A related issue arises if we set `data.x.y = null`. This should fail validation, as `data.x.y` is not a plain `object`. What happens, however, is that the validator throws, as you cannot read property `z` of `null`.

**Comment by kgryte** [comment]
Also, as a non-related issue, in looking at the above code, you could probably cut out all the `if (validate.errors === null) validate.errors = []` checks.

You could do so by setting `validate.errors` to an empty `array` at validation. As `validate.errors` would already be an `array`, no need to check.

Once validation has completed, you can check the length of `validate.errors`. If empty, set `validate.errors` equal to `null`.

**Comment by mafintosh** [comment]
@kgryte could you add a test case that fails?

**Comment by mafintosh** [comment]
this is fixed in 2.4.1. thanks for adding the test case and bringing this to my attention! :)

**Comment by kgryte** [comment]
@mafintosh Awesome. Thanks for fixing it!
